faucet/main.go

```diff
@@ -464,14 +476,74 @@ func prepareTokenMiddleware(redisClient *redis.Client) faucet.Middleware {
 				return
 			}
 
-			// Continue with serving the faucet request
-			next.ServeHTTP(w, r)
-
 			_, err = redisClient.HSet(r.Context(), "TOKEN:"+token, "used", "true").Result()
 			if err != nil {
 				http.Error(w, "Unable to lock token", http.StatusInternalServerError)
 				return
 			}
+
+			// Parse the request to extract the address
+			// XXX copied from prepareFundMiddleware, don't have time to refactor without unit tests!
+			var request faucet.Request
+			body, err := io.ReadAll(r.Body)
+			if err != nil {
+				http.Error(w, "Error reading request body", http.StatusInternalServerError)
+				return
+			}
+			// Close the original body
+			if err := r.Body.Close(); err != nil {
+				http.Error(w, "Error closing request body", http.StatusInternalServerError)
+				return
+			}
+			// Create a new ReadCloser from the read bytes
+			// so that future middleware will be able to read
+			r.Body = io.NopCloser(bytes.NewReader(body))
+			// Decode the original request
+			if err := json.NewDecoder(bytes.NewBuffer(body)).Decode(&request); err != nil {
+				http.Error(w, "Invalid request", http.StatusBadRequest)
+				return
+			}
+
+			// Read email
+			email, err := redisClient.HGet(r.Context(), "TOKEN:"+token, "email").Result()
+			if err != nil {
+				http.Error(w, "Unable to read token email", http.StatusInternalServerError)
+				return
+			}
+			// Store gno adress, email and token, so they are retrievable via the
+			// getEmail middleware.
+			err = redisClient.HSet(r.Context(), "GNO:"+request.To, "email", email, "token", token).Err()
+			if err != nil {
+				http.Error(w, "Unable to set gno address and email", http.StatusInternalServerError)
+				return
+			}
+
+			// Continue with serving the faucet request
+			next.ServeHTTP(w, r)
+		})
+	}
+}
+
+// getEmail detects the presence of `addr` form value, if present returns the
```

Comment:
I propose to put this in a lambda functions

Reply:
sounds good, I removed it [22f83fc](https://github.com/gnolang/gnochess/pull/153/commits/22f83fcdf86c40b6886e8ab2c02ede718d60d4ef)